### ISSUE TITLE
Fix S3966 FP: Enumerating a collection of tuples

### ISSUE THREAD

**Issue by odalet** [SonarSource/sonar-dotnet#7497]
### Description

When Iterating over a collection that returns a tuple, one field of which is IDisposable, and then disposing it inside the loop, a false positive S3966 is reported.

### Repro steps

In the example below, the first example does not report S3966 but we are iterating over a simple value. The 2 other examples report S3966 though:

```csharp
public sealed class Disposable : IDisposable
{
    private readonly int index;
    public Disposable(int i) => index = i;
    public void Dispose() => Console.WriteLine($"Disposing of {index}");
}

internal static class Program
{
    static void Main()
    {
        Console.WriteLine("Test #0 - OK (no FP");
        foreach (var disposable in GetDisposables())
            disposable.Dispose();

        Console.WriteLine("Test #1 - KO");
        foreach (var (_, disposable) in GetDisposableList())
            disposable.Dispose();

        Console.WriteLine();
        
        Console.WriteLine("Test #2 - KO");
        var enumerator = GetDisposableEnumerator();
        while (true)
        {
            if (!enumerator.MoveNext())
                break;

            var (_, disposable) = enumerator.Current;
            disposable.Dispose();
        }

        Console.WriteLine("Done");
    }

    private static IEnumerator<(int count, Disposable disposable)> GetDisposableEnumerator()
    {
        for (var i = 0; i < 10; i++)
            yield return (i, new(i));
    }

    private static List<(int count, Disposable disposable)> GetDisposableList() =>
        Enumerable.Range(0, 10).Select(i => (i, new Disposable(i))).ToList();

    private static List<Disposable> GetDisposables() =>
        Enumerable.Range(0, 10).Select(i => new Disposable(i)).ToList();
}
```

Please provide a minimal code snippet that compiles which reproduces the problem. Alternatively, you can share a link to the SonarCloud public project where the issue appears. If the issue requires some specific project configuration to reproduce, you can create a minimal reproducible project and share its git repository or attach it as a ZIP file.

### Expected behavior

Since the tuple member is not disposed of anywhere else, S3966 should not be raised

### Actual behavior

![image](https://github.com/SonarSource/sonar-dotnet/assets/79333/916c5702-328c-48f3-a268-264b0074b93f)

### Known workarounds

Using `SuppressMessageAttribute`...

### Related information

* C#/VB.NET Plugins version: **SonarAnalyzer.CSharp version 9.4.0.72892**
* Visual Studio version: **17.6.2**
* MSBuild / dotnet version: `net6.0` sdk-style project
* SonarScanner for .NET version (if used)
* Operating System: Windows 10 19045.3086 x64

**Comment by Tim-Pohlmann** [comment]
Hi @odalet, thanks for reporting it. I confirm this is an FP, and I will add this to our backlog. It will get fixed in a future hardening sprint.

**Comment by martin-strecker-sonarsource** [comment]
Tracked by https://sonarsource.atlassian.net/browse/NET-1203 now

**Comment by mary-georgiou-sonarsource** [comment]
Fixed in sonar-dotnet 10.9
